gtsfm/scene_optimizer.py

```diff
@@ -432,11 +443,62 @@ def save_full_frontend_metrics(
                 if report.reproj_error_gt_model is not None and report.v_corr_idxs_inlier_mask_gt is not None
                 else None,
                 "inlier_ratio_est_model": round_fn(report.inlier_ratio_est_model),
-                "num_inliers_est_model": report.num_inliers_est_model,
+                "num_inliers_est_model": int(report.num_inliers_est_model)
+                if report.num_inliers_est_model is not None
+                else None,
             }
         )
 
     io_utils.save_json_file(os.path.join(METRICS_PATH, filename), metrics_list)
 
     # Save duplicate copy of 'frontend_full.json' within React Folder.
     io_utils.save_json_file(os.path.join(REACT_METRICS_PATH, filename), metrics_list)
+
+    if matching_regime not in [ImageMatchingRegime.RETRIEVAL, ImageMatchingRegime.SEQUENTIAL_WITH_RETRIEVAL]:
+        return
+    if "VIEWGRAPH_2VIEW_REPORT" in filename:
+        # must come after two-view report file is written to disk in the Dask dependency graph.
+        _save_retrieval_two_view_metrics()
+
+
+def _save_retrieval_two_view_metrics() -> None:
+    """Compare 2-view similarity scores with their 2-view pose errors after viewgraph estimation."""
+    sim_fpath = os.path.join(PLOT_BASE_PATH, "netvlad_similarity_matrix.txt")
+    sim = np.loadtxt(sim_fpath, delimiter=",")
+
+    json_fpath = os.path.join(METRICS_PATH, "two_view_report_VIEWGRAPH_2VIEW_REPORT.json")
+    json_data = io_utils.read_json_file(json_fpath)
+
+    sim_scores = []
+    R_errors = []
+    U_errors = []
+
+    for entry in json_data:
+        i1 = entry["i1"]
+        i2 = entry["i2"]
+        R_error = entry["rotation_angular_error"]
+        U_error = entry["translation_angular_error"]
+        sim_score = sim[i1, i2]
+
+        sim_scores.append(sim_score)
+        R_errors.append(R_error)
+        U_errors.append(U_error)
+
+    plt.scatter(sim_scores, R_errors, 10, color="r", marker=".")
+    plt.xlabel("Similarity score")
+    plt.ylabel("Rotation error w.r.t. GT (deg.)")
+    plt.savefig(os.path.join(PLOT_BASE_PATH, "gt_rot_error_vs_similarity_score.jpg"), dpi=500)
+    plt.close("all")
```

Comment:
isnt this to be called only once?

Reply:
Hmm would need to be cleared after every plot call, to prevent aggregation on the same background plot.